### ISSUE TITLE
Link relations in HTTP Link header conflict with RFC 5988.

### ISSUE THREAD

**Issue by Zegnat** [indieweb/indieauth#17]
IndieAuth [registers](https://indieauth.spec.indieweb.org/#iana-considerations) 3 link relations: `authorization_endpoint`, `token_endpoint`, and `redirect_uri`. Additionally it requires both [clients](https://indieauth.spec.indieweb.org/#discovery-by-clients) and [servers](https://indieauth.spec.indieweb.org/#redirect-url) to check HTTP `Link` headers for URLs with these relations.

According to [RFC 5988](https://tools.ietf.org/html/rfc5988) (as cited by IndieAuth) an `_` (underscore) cannot be used by a link relation within [the header field](https://tools.ietf.org/html/rfc5988#section-5). A link’s relation type must match:

> ```
>   relation-type  = reg-rel-type | ext-rel-type
>   reg-rel-type   = LOALPHA *( LOALPHA | DIGIT | "." | "-" )
>   ext-rel-type   = URI
> ```

[RFC 8288](https://tools.ietf.org/html/rfc8288) (which replaces RFC 5988, cf. #15) does not solve this for us. There a link’s [relation type](https://tools.ietf.org/html/rfc8288#section-3) is defined using a slightly different ABNF but otherwise unchanged:

> ```
>   relation-type  = reg-rel-type / ext-rel-type
>   reg-rel-type   = LOALPHA *( LOALPHA / DIGIT / "." / "-" )
>   ext-rel-type   = URI ; Section 3 of [RFC3986]
> ```

**Comment by aaronpk** [comment]
I'm not sure the best path forward on this. We used underscores for `authorization_endpoint` because that's the same name as the property in [OpenID Connect Discovery](http://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata). 

It seems that parsers are not breaking on the underscore, so I would hate to make a breaking change in the spec just to stay conformant with RFC 5988/8288 when in reality the live code seems to not care.

**Comment by Zegnat** [comment]
> […] I would hate to make a breaking change in the spec just to stay conformant with RFC 5988/8288 when in reality the live code seems to not care.

I agree. So this becomes a question of whether the IndieAuth specification needs to call notice to this deviation from RFC 8288. Alternatively the issue gets ignored completely, or it is filed somewhere else.

---

As far as [link relations in HTML](https://html.spec.whatwg.org/multipage/links.html#other-link-types) are concerned, the WHATWG accepts any “[JavaScript string](https://infra.spec.whatwg.org/#javascript-string)” (basically: any valid UTF string) that does not contain [ASCII whitespace](https://infra.spec.whatwg.org/#ascii-whitespace). As long as the type is “[registered in the microformats wiki](https://html.spec.whatwg.org/multipage/links.html#other-link-types)”. Underscores are definitely fine there, and have some prior art.

**Comment by dshanske** [comment]
@Zegnat Do you consider this resolved as we used a - in the replacement for these, save redirect_uri ?

**Comment by Zegnat** [comment]
@dshanske Feels like this is pending closing by https://github.com/indieweb/indieauth/issues/133. As that would potentially introduce a different recommended way to detect `redirect_uri` (as `redirect_urls`) in the Client ID Document.
